### DISCUSSION TITLE
String Manipulation Functions

### DISCUSSION THREAD

**Discussion by springcomp** [jmespath-community/jmespath.spec#21] [JEP] [answered]
I think a `format` function would be great for maximum flexibility. However, I still think a core set of useful and simple functions is needed.

Here are string manipulation functions that are available in the programming languages that underlay the various popular JMESPath implementations:

| Function | Go | JavaScript | .NET | Python 
|---|---|---|---|---|
|[find_last](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3308144)|LastIndex|lastIndexOf|LastIndexOf|·
|[find_first](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3308144)|Index|indexOf|IndexOf|find
|format|_fmt.Sprintf()_||Format|format
|[lower](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3305907)|ToLower|toLowerCase|ToLower|lower
|[pad_left](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3306597)|·|padStart|PadLeft|rjust
|[pad_right](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3306597)|·|padEnd|PadRight|ljust
|[replace](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3306413)|Replace|replace|Replace|replace
|slice|(_intrinsic_)|slice|·|(_intrinsic_)
|[split](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3308432)|SplitN|split|Split|split
|title|Title|·|·|title
|[trim_left](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3306254)|TrimLeft|trimStart|TrimStart|lstrip
|[trim_right](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3306254)|TrimRight|trimEnd|TrimEnd|rstrip
|[trim](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3306254)|Trim|trim|Trim|strip
|[upper](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3305907)|ToUpper|toUpperCase|ToUpper|upper

**Note** that JMESPath already implements the following string manipulation functions:

- `contains`
- `ends_with`
- `join`
- `length` 
- `starts_with`
- `to_number`

We plan to propose [expanding slices to operate on string types](https://github.com/jmespath-community/jmespath.spec/discussions/26#discussioncomment-3284127).
This will enable a more powerful equivalent of a `substring` function.

Please, note that JMESPath already implements `reverse` on strings that overlaps with the `[::-1]` slice expression.

## Specification

Some string manipulation functions bring the new concept of _optional arguments_ to JMESPath functions. The specification paragraph on function evaluation must thus be changed accordingly – highlighted in **bold** in the text below:

_Functions can ~~either~~ have a specific arity, **a range of valid – minimum and maximum – number of arguments** or be variadic with a minimum number of arguments. If a function-expression is encountered where the arity does not match or the minimum number of arguments for a variadic function is not provided, then implementations must indicate to the caller than an invalid-arity error occurred. How and when this error is raised is implementation specific._

## References

- [Go strings package documentation](https://pkg.go.dev/strings)
- [Properties of the JavaScript String Prototype Object](https://262.ecma-international.org/13.0/#sec-properties-of-the-string-prototype-object)
- [C# String class](https://docs.microsoft.com/en-us/dotnet/api/system.string?view=net-6.0)
- [Python String methods](https://docs.python.org/3/library/stdtypes.html#string-methods)

**Comment by innovate-invent** [comment]
Some additional functions:

| | |
|-|-|
trim_left(string $subject, string\|array[string] $prefix) | Trim $prefix or any string in $prefix from left of $subject |
trim_right(string $subject, string\|array[string] $suffix) | Trim $suffix or any string in $suffix from right of $subject |
"abc"[::-1] | Reverse string |
"abc"[1:] | Substring |
pad_left(string $subject, number $width, string $char = " ") | Pad left of $subject with $char to $width |
pad_right(string $subject, number $width, string $char = " ") | Pad right of $subject with $char to $width |
upper(string $subject) | Set all characters in $subject to uppercase |
lower(string $subject) | Set all characters in $subject to lowercase |
title(string $subject) | Change $subject to Title Case |

An alternative to pad_*, upper, lower, and title could be a format() function that takes a format string with various formatting codes / syntax that include these manipulations.

I think the [Python format string syntax](https://docs.python.org/3/library/string.html#format-string-syntax) is worth adopting. It allows for parametrization of the placeholders within the format string. This opens the doors to extensibility over creating additional string manipulation functions..

**Reply by springcomp** [reply]
https://peps.python.org/pep-3101/

**Reply by springcomp** [reply]
String interpolation is indeed a powerful feature in many modern languages.
See #90.

**Comment by springcomp** [comment]
_This is a JEP fragment that specifies the `upper()` and `lower()` functions. Each JEP fragment has its own spaces for comments and feedback. Once sufficiently fleshed out, they will be merged into a single JEP_

[back to top](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussion-3914002)

## Specification

### lower

```
string lower(string $subject)
```
Returns the lowercase `$subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"STRING"` | `` lower(@) `` |  `"string"`

### upper

```
string upper(string $subject)
```
Returns the uppercase `$subject` string.

| Given | Expression | Result
|---|---|---
| `"string"` | `` upper(@) `` |  `"STRING"`

## History

N/A

**Comment by springcomp** [comment]
_This is a JEP fragment that specifies the `trim()`, `trim_start()` and `trim_end()` functions. Each JEP fragment has its own spaces for comments and feedback. Once sufficiently fleshed out, they will be merged into a single JEP_

[back to top](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussion-3914002)

## Specification

### trim

```
string trim(string $subject[, string $chars])
```
Given the `$subject` string, `trim()` removes the leading and trailing characters found in `$chars`.

The `$chars` optional string parameter represents a set of characters to be removed. If this parameter is not specified, or is an empty string, whitespace characters are removed from the `$subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"   subject string   "` | `` trim(@) `` |  `"subject string"`
| `"   subject string   "` | `` trim(@, '') `` |  `"subject string"`
| `"   subject string   "` | `` trim(@, ' ') `` |  `"subject string"`
| `"   subject string   "` | `` trim(@, 's') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim(@, 'su') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim(@, 'su ') `` |  `"bject string"`
| `"   subject string   "` | `` trim(@, 'gsu ') `` |  `"bject strin"`

### trim_left

```
string trim_left(string $subject[, string $chars])
```
Given the `$subject` string, `trim_left()` removes the leading characters found in `$chars`.

Like for the `trim()` function, the `$chars` optional string parameter represents a set of characters to be removed. `trim_left()` defaults to removing whitespace characters if `$chars` is not specified or is and empty string.

### Examples

| Given | Expression | Result
|---|---|---
| `"   subject string   "` | `` trim_left(@) `` |  `"subject string   "`
| `"   subject string   "` | `` trim_left(@, 's') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_left(@, 'su') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_left(@, 'su ') `` |  `"bject string"`
| `"   subject string   "` | `` trim_left(@, 'gsu ') `` |  `"bject string"`

### trim_right

```
string trim_right(string $subject[, string $chars])
```
Given the `$subject` string, `trim_right()` removes the trailing characters found in `$chars`.

Like for the `trim()` and `trim_left()` functions, the `$chars` optional string parameter represents a set of characters to be removed. `trim_right()` defaults to removing whitespace characters if `$chars` is not specified or is and empty string.

### Examples

| Given | Expression | Result
|---|---|---
| `"   subject string   "` | `` trim_right(@) `` |  `"   subject string"`
| `"   subject string   "` | `` trim_right(@, 's') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_right(@, 'su') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_right(@, 'su ') `` |  `"   subject string"`
| `"   subject string   "` | `` trim_right(@, 'gsu ') `` |  `"   subject strin"`


## History

N/A

**Comment by springcomp** [comment]
_This is a JEP fragment that specifies the `replace()` function. Each JEP fragment has its own spaces for comments and feedback. Once sufficiently fleshed out, they will be merged into a single JEP_

[back to top](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussion-3914002)

## Specification

### replace

```
string replace(string $subject, string $old, string $new[, number $count])
```
Given the `$subject` string, `replace()` replaces all occurrences of the `$old` substring with the `$new` substring.

The `$count` optional parameter specifies how many occurrences of the `$old` substring in `$subject` are replaced. If this parameter is omitted, all occurrences are replaced. If `$count` is negative, an error MUST be raised.

The `replace()` function has no effect if `$count` is `0`.

### Examples

| Given | Expression | Result
|---|---|---
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `0`) `` |  `"aabaaabaaaab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `1`) `` |  `"-baaabaaaab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `2`) `` |  `"-b-abaaaab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `3`) `` |  `"-b-ab-aab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-') `` |  `"-b-ab--b"`

## History

N/A

**Comment by springcomp** [comment]
_This is a JEP fragment that specifies the `pad_left()` and `pad_right()` functions. Each JEP fragment has its own spaces for comments and feedback. Once sufficiently fleshed out, they will be merged into a single JEP_

[back to top](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussion-3914002)

## Specification

### pad_left

```
string pad_left(string $subject, number $width[, string $pad])
```

Given the `$subject` string, `pad_left()` adds `$width - length($subject)` characters to the beginning of the string and returns a string of length `$width`.

The `$pad` optional string parameter specifies the padding character to use. If this parameter is omitted, `pad_left()` adds one of more ASCII space characters to the beginning of the string. If present, the `$pad` string must have a single character, otherwise, an error MUST be raised.

The `pad_left()` function has no effect if `$count` is less than, or equal to the length of the ` $subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"string"` | `` pad_left(@, `0`) `` |  `"string"`
| `"string"` | `` pad_left(@, `5`) `` |  `"string"`
| `"string"` | `` pad_left(@, `10`) `` |  `"    string"`
| `"string"` | `` pad_left(@, `10`, '-') `` |  `"----string"`

### pad_right

```
string pad_right(string $subject, number $width[, string $pad])
```

Given the `$subject` string, `pad_right()` adds `$width - length($subject)` characters to the end of the string and returns a string of length `$width`.

The `$pad` optional string parameter specifies the padding character to use. If this parameter is omitted, `pad_right()` adds ASCII space characters to the end of the string. If present, the `$pad` string must have a single character, otherwise, an error MUST be raised.

The `pad_right()` function has no effect if `$count` is less than, or equal to the length of the ` $subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"string"` | `` pad_right(@, `0`) `` |  `"string"`
| `"string"` | `` pad_right(@, `5`) `` |  `"string"`
| `"string"` | `` pad_right(@, `10`) `` |  `"string    "`
| `"string"` | `` pad_right(@, `10`, '-') `` |  `"string----"`

## History

|Date|Change|
|---|---|
|15-August-2022|Renamed `$count` argument to the more meaningful  `$width`.

**Comment by springcomp** [comment]
_This is a JEP fragment that specifies the `find_first()`, `find_last()` functions. Each JEP fragment has its own spaces for comments and feedback. Once sufficiently fleshed out, they will be merged into a single JEP_

[back to top](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussion-3914002)

## Specification

### find_first

```
string find_first(string $subject, string $sub[, int $start[, int $end]])
```
Given the `$subject` string, `find_first()` returns the index of the first occurence where the `$sub` substring appears in `$subject` or `null`.

The `$start` and `$end` parameters are optional and allow to select where `find_first()` must perform its search within `$subject`.

- If `$start` is omitted, it defaults to `0` which is the start of the `$subject` string.
- If `$end` is omitted, it defaults to `length($subject) -1` which is the end of the string.

### Examples

| Given | Expression | Result
|---|---|---
| `"subject string"` | `` find_first(@, 'string') `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `0`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `0`, `14`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `-99`, `100`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `0`, `13`) `` |  `null`
| `"subject string"` | `` find_first(@, 'string', `8`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `9`) `` |  `null`
| `"subject string"` | `` find_first(@, 's') `` |  `0`
| `"subject string"` | `` find_first(@, 's', 1) `` |  `8`

### find_last

```
string find_last(string $subject, string $sub[, int $pos])
```
Given the `$subject` string, `find_last()` returns the index of the last occurence where the `$sub` substring appears in `$subject` or `null`.

The `$pos` optional parameter allow to select where `find_last()` must perform its search within `$subject`.

If this parameter is omitted, It defaults to `length($subject) - 1` which is the end of the `$subject` string. `find_last()` then searches backwards in the `$subject` string for the first occurrence of `$sub`.

| Given | Expression | Result
|---|---|---
| `"subject string"` | `` find_last(@, 'string') `` |  `8`
| `"subject string"` | `` find_last(@, 'string', 8) `` |  `8`
| `"subject string"` | `` find_last(@, 'string', 7) `` |  `null`
| `"subject string"` | `` find_last(@, 's', 8) `` |  `8`
| `"subject string"` | `` find_last(@, 's', 7) `` |  `0`


## History

N/A

**Comment by springcomp** [comment]
_This is a JEP fragment that specifies the `split()` function as specified by Chris Armstrong from Gorilla Stack. Each JEP fragment has its own spaces for comments and feedback. Once sufficiently fleshed out, they will be merged into a single JEP_

[back to top](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussion-3914002)

## Specification

### split

```
array[string] split(string $subject, string $search[, number $count])
```

Given string `$subject`, `split()` breaks on occurrences of the string `$search` and returns an array.

The `split()` function returns an array containing each partial string between occurrences of `$search`. If  `$subject` contains no occurrences of the `$search` string, an array containing just the original `$subject` string will be returned.

The `$count` optional parameter specifies how many occurrences of the `$search` string are split. If this parameter is omitted, all occurrences are split. If `$count` is negative, and error MUST be raise.

If `$count` is equal to `0`, `split()` returns an array containing a single element, the `$subject` string.

Otherwise, the `split()` function breaks on occurrences of the `$search` string up to `$count` times. The last string in the resulting array containing the remaining contents of `$subject` unmodified.

### Examples

| Expression | Result 
|---|---
| `split('all chars', '')` | `[ "a", "l", "l", " ", "c", "h", "a", "r", "s" ]`
| `split('/', '/')` | `[ "", "" ]` |
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|')` | `[ "average", "min", "max", "mean", "median" ]` 
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', 3)` | `[ "average", "min", "max", "mean\|-\|median" ]`
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', 2)` | `[ "average", "min", "max\|-\|mean\|-\|median" ]`
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', 1)` | `[ "average", "min\|-\|max\|-\|mean\|-\|median" ]`
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', 0)` | `[ "average\|-\|min\|-\|max\|-\|mean\|-\|median" ]`
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '-')` | `[ "average\|", "\|min\|", "\|max\|", "\|mean\|", "\|median" ]`

## History

|Change|Description
|---|---|
|15-August-2022|Reworked document structure.
|15-August-2022|Renamed optional argument `$max` to `$count` and removed irrelevant default value `-1`.
|15-August-2022|Renamed argument `$split_str` to `$search`.
|15-August-2022|`split()` splits up to `$count` times, rather than the original `$count – 1` times.

**Reply by springcomp** [reply]
Seems there is an inconsistency with the `$max` parameter (**Edit**: renamed to `$count`).

Based on the original examples reproduced here:

| Expression | Result 
|---|---
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|mode\|-\|median', '\|-\|')` | `["average", "min", "max", "mean", "mode", "median"]`|
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|mode\|-\|median', '\|-\|', 3)` | `["average", "min", "max\|-\|mean\|-\|mode\|-\|median"]`
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|mode\|-\|median', '\|-\|', 0)` | `["average\|-\|min\|-\|max\|-\|mean\|-\|mode\|-\|median"]`

- When `$max` is `3`, the resulting array splits up to `2` occurrences.
- When `$max` is `0`, the resulting array splits up to `0` occurrences.

This seems to me we should generalize:

- When `$max` is `2`, the resulting array splits up to `1` occurrences.
- When `$max` is `1`, the resulting array splits up to `0` occurrences.

Thus, we arrive at an inconsistency, with the same result being returned when `$max` is `0` or `1`.

I think, for the sake of consistency, we should treat the `$max` as  a `$count` parameter instead, similar to the argument with the same name in [`replace()`](https://github.com/jmespath-community/jmespath.spec/discussions/21#discussioncomment-3306413) function.

This is consistent with the Python and JavaScript implementation. However, this conflicts with the .NET implementation.

**Reply by springcomp** [reply]
Should we add a `[, bool $removeEmptyEnties]` optional parameter?

The first motivating example would then read, with or without this parameter:

- `` search( split('/', '/'), {} ) -> ["", ""] ``
- `` search( split('/', '/', `null`, `true`), {} ) -> []``

In that case, the function evaluation specification text must be changed to support specifying some, but not all, of the optional parameters. Maybe adding text like:

_**To call a function with some, but not all, of the optional parameters, unspecified parameters can be designated by passing a `null` value instead. Thus, optional parameters can either be omitted, or specified explicitly with a `null` value.**_

**Reply by springcomp** [reply]
> Should we add a `[, bool $removeEmptyEnties]` optional parameter?

This is not really needed after all.
Because empty strings are _falsy_, we can remove empty entries using the following expression:

- `` search( [?@], ["", ""] ) -> [] ``

**Comment by springcomp** [comment]
The string manipulation functions are supported in the [preview page](https://jmespath.site/#preview).
Please, give them a try!

**Comment by springcomp** [comment, marked as answer]
# String Functions

|||
|---|---
| **JEP**    |  14
| **Author** | Maxime Labelle, Chris Armstrong (GorillaStack)
| **Created**| 13-October-2022
| **SemVer** | MINOR
| **Status**| Draft
| **[Discussion #21]** | #21

## Abstract

This JEP introduces a core set of useful string manipulation functions. Those functions are modeled from functions found in popular programming languages such as JavaScript and Python.

## Specification

### find_first

```
int find_first(string $subject, string $sub[, int $start[, int $end]])
```
Given the `$subject` string, `find_first()` returns the index of the first occurence where the `$sub` substring appears in `$subject` or `null`.

The `$start` and `$end` parameters are optional and allow to select where `find_first()` must perform its search within `$subject`.

- If `$start` is omitted, it defaults to `0` which is the start of the `$subject` string.
- If `$end` is omitted, it defaults to `length($subject) - 1` which is the end of the string.

Contrary to similar functions found in most popular programming languages, the `find_first()` function does not return `-1` if the occurrence of the `$sub` substring cannot be found. Instead, it returns `null` for consistency reasons with how JMESPath behaves. 

### Examples

| Given | Expression | Result
|---|---|---
| `"subject string"` | `` find_first(@, 'string') `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `0`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `0`, `14`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `-99`, `100`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `0`, `13`) `` |  `null`
| `"subject string"` | `` find_first(@, 'string', `8`) `` |  `8`
| `"subject string"` | `` find_first(@, 'string', `9`) `` |  `null`
| `"subject string"` | `` find_first(@, 's') `` |  `0`
| `"subject string"` | `` find_first(@, 's', 1) `` |  `8`

### find_last

```
int find_last(string $subject, string $sub[, int $pos])
```
Given the `$subject` string, `find_last()` returns the index of the last occurence where the `$sub` substring appears in `$subject` or `null`.

The `$pos` optional parameter allow to select where `find_last()` must perform its search within `$subject`.

If this parameter is omitted, It defaults to `length($subject) - 1` which is the end of the `$subject` string. `find_last()` then searches backwards in the `$subject` string for the first occurrence of `$sub`.

Likewise, the `find_last()` function does not return `-1` if the occurrence of the `$sub` substring cannot be found. Instead, it returns `null` for consistency reasons with how JMESPath behaves. 

| Given | Expression | Result
|---|---|---
| `"subject string"` | `` find_last(@, 'string') `` |  `8`
| `"subject string"` | `` find_last(@, 'string', 8) `` |  `8`
| `"subject string"` | `` find_last(@, 'string', 7) `` |  `null`
| `"subject string"` | `` find_last(@, 's', 8) `` |  `8`
| `"subject string"` | `` find_last(@, 's', 7) `` |  `0`

### lower

```
string lower(string $subject)
```
Returns the lowercase `$subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"STRING"` | `` lower(@) `` |  `"string"`

### pad_left

```
string pad_left(string $subject, number $width[, string $pad])
```

Given the `$subject` string, `pad_left()` adds `$width - length($subject)` characters to the beginning of the string and returns a string of length `$width`.

The `$pad` optional string parameter specifies the padding character to use. If this parameter is omitted, `pad_left()` adds one of more ASCII space characters to the beginning of the string. If present, the `$pad` string must have a single character, otherwise, an error MUST be raised.

The `pad_left()` function has no effect if `$width` is less than, or equal to the length of the ` $subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"string"` | `` pad_left(@, `0`) `` |  `"string"`
| `"string"` | `` pad_left(@, `5`) `` |  `"string"`
| `"string"` | `` pad_left(@, `10`) `` |  `"    string"`
| `"string"` | `` pad_left(@, `10`, '-') `` |  `"----string"`

### pad_right

```
string pad_right(string $subject, number $width[, string $pad])
```

Given the `$subject` string, `pad_right()` adds `$width - length($subject)` characters to the end of the string and returns a string of length `$width`.

The `$pad` optional string parameter specifies the padding character to use. If this parameter is omitted, `pad_right()` adds ASCII space characters to the end of the string. If present, the `$pad` string must have a single character, otherwise, an error MUST be raised.

The `pad_right()` function has no effect if `$width` is less than, or equal to the length of the ` $subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"string"` | `` pad_right(@, `0`) `` |  `"string"`
| `"string"` | `` pad_right(@, `5`) `` |  `"string"`
| `"string"` | `` pad_right(@, `10`) `` |  `"string    "`
| `"string"` | `` pad_right(@, `10`, '-') `` |  `"string----"`

### replace

```
string replace(string $subject, string $old, string $new[, number $count])
```
Given the `$subject` string, `replace()` replaces all occurrences of the `$old` substring with the `$new` substring.

The `$count` optional parameter specifies how many occurrences of the `$old` substring in `$subject` are replaced. If this parameter is omitted, all occurrences are replaced. If `$count` is negative, an error MUST be raised.

The `replace()` function has no effect if `$count` is `0`.

### Examples

| Given | Expression | Result
|---|---|---
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `0`) `` |  `"aabaaabaaaab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `1`) `` |  `"-baaabaaaab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `2`) `` |  `"-b-abaaaab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-', `3`) `` |  `"-b-ab-aab"`
| `"aabaaabaaaab"` | `` replace(@, 'aa', '-') `` |  `"-b-ab--b"`

### split

```
array[string] split(string $subject, string $search[, number $count])
```

Given string `$subject`, `split()` breaks on occurrences of the string `$search` and returns an array.

The `split()` function returns an array containing each partial string between occurrences of `$search`. If  `$subject` contains no occurrences of the `$search` string, an array containing just the original `$subject` string will be returned.

The `$count` optional parameter specifies how many occurrences of the `$search` string are split. If this parameter is omitted, all occurrences are split. If `$count` is negative, and error MUST be raised.

If `$count` is equal to `0`, `split()` returns an array containing a single element, the `$subject` string.

Otherwise, the `split()` function breaks on occurrences of the `$search` string up to `$count` times. The last string in the resulting array containing the remaining contents of `$subject` unmodified.

**Note**: The `split()` function was [originally designed by Chris Armstrong](https://github.com/GorillaStack/jmespath.site/blob/master/docs/proposals/string-manipulation.rst). However, its behaviour has been slightly altered for consistency reasons.

### Examples

| Expression | Result 
|---|---
| `split('all chars', '')` | `[ "a", "l", "l", " ", "c", "h", "a", "r", "s" ]`
| `split('/', '/')` | `[ "", "" ]` |
|`` split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|') `` | `[ "average", "min", "max", "mean", "median" ]` 
|`` split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', `3`) `` | `[ "average", "min", "max", "mean\|-\|median" ]`
|`` split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', `2`) `` | `[ "average", "min", "max\|-\|mean\|-\|median" ]`
|`` split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', `1`) `` | `[ "average", "min\|-\|max\|-\|mean\|-\|median" ]`
|`` split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '\|-\|', `0`) `` | `[ "average\|-\|min\|-\|max\|-\|mean\|-\|median" ]`
| `split('average\|-\|min\|-\|max\|-\|mean\|-\|median', '-')` | `[ "average\|", "\|min\|", "\|max\|", "\|mean\|", "\|median" ]`

## Specification

### trim

```
string trim(string $subject[, string $chars])
```
Given the `$subject` string, `trim()` removes the leading and trailing characters found in `$chars`.

The `$chars` optional string parameter represents a set of characters to be removed. If this parameter is not specified, or is an empty string, whitespace characters are removed from the `$subject` string.

### Examples

| Given | Expression | Result
|---|---|---
| `"   subject string   "` | `` trim(@) `` |  `"subject string"`
| `"   subject string   "` | `` trim(@, '') `` |  `"subject string"`
| `"   subject string   "` | `` trim(@, ' ') `` |  `"subject string"`
| `"   subject string   "` | `` trim(@, 's') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim(@, 'su') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim(@, 'su ') `` |  `"bject string"`
| `"   subject string   "` | `` trim(@, 'gsu ') `` |  `"bject strin"`

### trim_left

```
string trim_left(string $subject[, string $chars])
```
Given the `$subject` string, `trim_left()` removes the leading characters found in `$chars`.

Like for the `trim()` function, the `$chars` optional string parameter represents a set of characters to be removed. `trim_left()` defaults to removing whitespace characters if `$chars` is not specified or is an empty string.

### Examples

| Given | Expression | Result
|---|---|---
| `"   subject string   "` | `` trim_left(@) `` |  `"subject string   "`
| `"   subject string   "` | `` trim_left(@, 's') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_left(@, 'su') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_left(@, 'su ') `` |  `"bject string   "`
| `"   subject string   "` | `` trim_left(@, 'gsu ') `` |  `"bject string   "`

### trim_right

```
string trim_right(string $subject[, string $chars])
```
Given the `$subject` string, `trim_right()` removes the trailing characters found in `$chars`.

Like for the `trim()` and `trim_left()` functions, the `$chars` optional string parameter represents a set of characters to be removed. `trim_right()` defaults to removing whitespace characters if `$chars` is not specified or is an empty string.

### Examples

| Given | Expression | Result
|---|---|---
| `"   subject string   "` | `` trim_right(@) `` |  `"   subject string"`
| `"   subject string   "` | `` trim_right(@, 's') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_right(@, 'su') `` |  `"   subject string   "`
| `"   subject string   "` | `` trim_right(@, 'su ') `` |  `"   subject string"`
| `"   subject string   "` | `` trim_right(@, 'gsu ') `` |  `"   subject strin"`

### upper

```
string upper(string $subject)
```
Returns the uppercase `$subject` string.

| Given | Expression | Result
|---|---|---
| `"string"` | `` upper(@) `` |  `"STRING"`

## Compliance

A new `string_functions.json` file will be added to the compliance tests.

**Comment by hktonylee** [comment]
Thanks for the proposal @springcomp. Can we also add a `case_fold` for a proper case insensitive comparison (in additional to `lower` / `upper`)?

For example, the lower case of `ß` is `ß`. The lower case of `ss` is `ss`. They are not the same if converting to lower case first. But they are same (case-insensitively) in German.

```js
// this is Javascript
'ß'.toLowerCase() == 'ss'.toLowerCase() // this is false
'ß'.localeCompare('ss', undefined, { sensitivity: 'base' })  // this is 0, they are the same with case insensitive comparison
```

**Reply by springcomp** [reply]
@hktonylee thanks for the feedback.

As a general principle, we try to not dwelve too deep in Unicode-specific issues within JMESPath as we think this is best handled at an application level. That’s why I’m weary to standardize such a function.

That said, that is a really good idea for a Unicode-specific set of extensions that could be provided alongside this project.

**Comment by vivek-1689** [comment]
Hi,
When can we expect this feature to be available for ruby version?

**Reply by springcomp** [reply]
As soon as a sufficiently motivated contributor works on a fork to the ruby implementation 😁.

I ramped up my skills and implemented the [Python version](https://github.com/jmespath-community/python-jmespath).
I have [learned Go in 12 minutes](https://youtu.be/C8LgvuEBraI) and forked and implemented the [Golang version](https://github.com/jmespath-community/go-jmespath).
I have somehow pretended knowing TypeScript and implemented the [TypeScript version](https://github.com/jmespath-community/typescript-jmespath).

I’m running out of momentum, to be honest 🤷‍♂️ and do not see Ruby as worthy enough to learn in my skill set to tackle this myself.

**Comment by dust63** [comment]
I propose to add a regex operator, to allow to use the power of regex:

### Match operator
Given the `$subject` string, `regex_match()` find the `pattern`  and return a boolean result.

| Given | Expression | Result |
| --- | --- | --- |
| "foo" | regex_match(@, 'foo') | True |
| "foo" | regex_match(@, 'bar') | False|
| "foo.bar" | regex_match(@, '\.bar$') | True |
| "foo.oof" | regex_match(@, '\.bar$') | False |
| "johndoe@eurocorp.com" | regex_match(@, '^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$') | True |
| "johndoe.eurocorp.com" | regex_match(@, '^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$') | False |

**Reply by springcomp** [reply]
Thanks for your feedback.
There is [ongoing discussion](https://github.com/jmespath-community/jmespath.spec/discussions/23#discussioncomment-2290775) around this topic. Can I kindly request you to leave your feedback there ?

Also, there is a [pending proposal](https://github.com/jmespath/jmespath.jep/issues/29) submitted to the official specification but that is subject to the [interoperable regex RFC](https://www.rfc-editor.org/rfc/rfc9485) being approved. Please, note that we will be implementing that proposal as functions, pretty much in the same way as you are showing in your table.
